Sources/System/FilePath/FilePathWindows.swift

```diff
@@ -461,3 +461,46 @@ extension SystemString {
     return lexer.current
   }
 }
+
+#if os(Windows)
+import WinSDK
+
+extension UnsafePointer where Pointee == CInterop.PlatformChar {
+  /// Invokes `body` with a resolved and potentially `\\?\`-prefixed version of the pointee,
+  /// to ensure long paths greater than 260 characters are handled correctly.
+  ///
+  /// - seealso: https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation
+  internal func withCanonicalPathRepresentation<Result>(_ body: (Self) throws -> Result) throws -> Result {
+    // 1. Normalize the path first.
+    // Contrary to the documentation, this works on long paths independently
+    // of the registry or process setting to enable long paths (but it will also
+    // not add the \\?\ prefix required by other functions under these conditions).
+    let dwLength: DWORD = GetFullPathNameW(self, 0, nil, nil)
+    return try withUnsafeTemporaryAllocation(of: WCHAR.self, capacity: Int(dwLength)) { fullPath in
+      guard GetFullPathNameW(self, DWORD(fullPath.count), fullPath.baseAddress, nil) > 0 else {
+        throw Errno(rawValue: _mapWindowsErrorToErrno(GetLastError()))
+      }
+
+      // 1.5 Leave \\.\ prefixed paths alone since device paths are already an exact representation and PathCchCanonicalizeEx will mangle these.
+      if let base = fullPath.baseAddress, base[0] == UInt8(ascii: "\\"), base[1] == UInt8(ascii: "\\"), base[2] == UInt8(ascii: "."), base[3] == UInt8(ascii: "\\") {
+        return try body(base)
+      }
+
+      // 2. Canonicalize the path.
+      // This will add the \\?\ prefix if needed based on the path's length.
+      let capacity = Int16.max
+      return try withUnsafeTemporaryAllocation(of: WCHAR.self, capacity: numericCast(capacity)) { outBuffer in
```

Comment:
This is allocating 32767 bytes, and I expect that will be executed as a heap allocation, repeated every time this is called. Usually swift-system tries to avoid doing implicit allocations like that, as clients sometimes get infuriated when they find out that a low-level syscall wrapper library engages in those.

We may have already lost those clients on Windows; I worry that this may make it harder to try winning them back.

Reply:
I changed this to use PathAllocCanonicalize, which has Windows allocate the buffer and can theoretically be smaller, but it's always going to be a heap allocation.

That said, Win32 APIs are not syscalls; they are higher level than the Unix "equivalents". Windows does not have a stable syscall interface and thus it's undocumented. So I think we can get away with some heap allocations where necessary to use the APIs correctly, since many Win32 functions are already doing so internally.

Reply:
I _think_ the length of the result of the `Path*Canonicalize*` functions is always less than or equal to the length of the input string plus one (plus two if you've got a trailing `NUL`), so we don't need to allocate 32KB here.

I also wonder whether we could have a canonicalize-in-place function that mutates the underlying `SystemString` as a more efficient option.